### ISSUE TITLE
CLIEngine addPlugin function not working

### ISSUE THREAD

**Issue by stevenzeck** [eslint/eslint#11964]
**Tell us about your environment**

* **ESLint Version:** 6.0.1
* **Node Version:** 12.4.0
* **npm Version:** 6.9.0

**What parser (default, Babel-ESLint, etc.) are you using?**
Default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
{
    "parserOptions": {
        "ecmaFeatures": {
            "jsx": true
        }
    },
    "globals": {
        "createReactClass": true
    },
    "plugins": ["react"],
    "extends": ["plugin:react/recommended"],
    "rules": {
        "react/jsx-wrap-multilines": [
            "error", {
                "declaration": "parens-new-line",
                "assignment": "parens-new-line",
                "return": "parens-new-line",
                "arrow": "parens-new-line",
                "condition": "parens-new-line",
                "logical": "parens-new-line",
                "prop": "parens-new-line"
            }
        ]
    }
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

I'm trying to merge a PR that updates our ESLint dependency from 5.x to 6.x.  However, a test that adds react-eslint-plugin and formats React does not pass.  If I add it via the `plugins` string array in the CLIEngine options it runs.  PR I'm referring to is:  https://github.com/Unibeautify/beautifier-eslint/pull/142.

I tried debugging, and the best I can tell `additionalPluginPool` isn't used in the relevant functions (mainly executeOnText()).

```js
      const cliOptions: CLIOptions = {
        baseConfig: {
          extends: extendsToAdd,
          globals: globals,
        },
        fix: true,
        parserOptions: parseOpts,
        rules: finalOptions,
        useEslintrc: false,
      };
      const cli: CLIEngine = new CLIEngine(cliOptions);
      cli.addPlugin("react", reactPlugin.package);
      const report: LintReport = cli.executeOnText(text);
      const result: LintResult = report.results[0];
```

```bash

```

**What did you expect to happen?**
The plugin be used when added via addPlugin

**What actually happened? Please include the actual, raw output from ESLint.**
I don't see it using the plugin.

**Are you willing to submit a pull request to fix this bug?**
If I need to

**Comment by mysticatea** [comment]
Thank you for your report.

What do you use the config?
If your config doesn't have `plugins: ["react"]`, the config cannot use `react/*` rules.

**Comment by stevenzeck** [comment]
@mysticatea Sorry I missed that in my report, I updated the original comment with the configuration.  Note that the way it works is we grab the settings from the config via CLIEngine's [`getConfigForFile`](https://github.com/Unibeautify/beautifier-eslint/blob/master/src/index.ts#L44) and use that to build the CLIEngine [config](https://github.com/Unibeautify/beautifier-eslint/blob/master/src/index.ts#L74-L78).

It was working in 5.x, but the PR to update to 6.x doesn't work.  I can get the exact message list from executeOnText's report later today, but it was something along the lines of no rule found for react/*.

**Comment by jehy** [comment]
Same for me. I had a very simple plugin to count files:

```

  cli.addPlugin('counter', {
    processors: {
      '.js': {
        preprocess(text, filename) {
          counter++;
          if (showProgress)
          {
            progressBar.update(counter, {
              currentFile: filename.replace(myPath, ''),
            });
          }
          return [text];
        },
        postprocess(messages) {
          return messages[0];
        },
      },


    },
  });
```

And it was broken exactly in v6.0.0-alpha.1

I can create a separate issue for this but the problem seems the same. Please take a look, @mysticatea .

**Comment by platinumazure** [comment]
@stevenzeck What happens if you specify `"plugins": ["react"]` in the CLIEngine constructor options? From what I can tell, that should automatically load the plugins.

**Comment by platinumazure** [comment]
@jehy Could you please create a new issue? Thanks!

**Comment by stevenzeck** [comment]
@platinumazure yes it works when in the constructor. But that's not how were using it and prefer to use addPlugin.

**Comment by mysticatea** [comment]
OK. So it was a bug and ESLint 6 has fixed it.

The `plugins` setting is needed to use plugins. But previously, ESLint had a bug that we could use plugins without the `plugins` setting if the `CLIEngine` instance had loaded the plugins in the past.

**Comment by platinumazure** [comment]
@mysticatea Does that mean we no longer support dynamically adding plugins via `addPlugin`?

I don't think that was intended. To me, `"plugins"` option is for convenience and calls `addPlugin` behind the scenes.

**Comment by mysticatea** [comment]
`CLIEngine` still supports dynamically adding plugins via `addPlugin`. If a user did `engine.addPlugin("foo", foo)`, `CLIEngine` uses it on `plugins: ["foo"]` even if `node_modules/eslint-plugin-foo` didn't exist.

**Comment by mysticatea** [comment]
The `addPlugin()` method make the engine instance to use `eslint-plugin-xxx` from the added one instead of `node_modules/eslint-plugin-xxx`. But it *doesn't* let us be able to use plugin rules without `plugins` setting.

```yml
# .eslintrc.yml
root: true
rules:
    # Needs `plugins:["foo"]` to use `foo/a-rule`.
    foo/a-rule: error
```

However, previously, we could do such because of a bug. We were able to use `foo/a-rule` without `plugins` setting unintentionally if `eslint-plugin-foo` already existed in the loaded plugin cache.

**Comment by stevenzeck** [comment]
@mysticatea What about `addPlugin()` though?  That's what this issue is for, since it no longer works in ESLint 6.x.

**Comment by mysticatea** [comment]
@stevenzeck 
> The `addPlugin()` method make the engine instance to use `eslint-plugin-xxx` from the added one instead of `node_modules/eslint-plugin-xxx`. But it *doesn't* let us be able to use plugin rules without plugins setting.
> https://github.com/eslint/eslint/issues/11964#issuecomment-514908078

**Comment by stevenzeck** [comment]
@mysticatea Are you saying that you have to add the plugin to the `plugins` property in CLIOptions when creating the CLIEngine, even though we are it via `addPlugin()`?

In the repo I support, we're including the plugins' rules in CLIOptions' rules property, then we create the CLIEngine, then we add the plugin.  We don't include the `plugins` property in CLIOptions because we are adding it with `addPlugin()`.  Is that not supported anymore?

**Comment by mysticatea** [comment]
> Are you saying that you have to add the plugin to the plugins property in CLIOptions when creating the CLIEngine, even though we are it via addPlugin()?

Yes. The functionality of the `addPlugin()` method is to use the added plugins instead of `node_modules/eslint-plugin-xxx` on the `plugins` property in configs. If you didn't have the `plugins` property in your config, `addPlugin()` doesn't have any effects.

However, old ESLint had a bug that could use plugin rules without the `plugins` property if the plugin had been loaded in the past. The plugins that were added by `addPlugin()` had been addressed as such a loaded plugins.

**Comment by jehy** [comment]
Sounds like a really strange breaking change.
I suppose, adding a new api like "replacePlugin" would be fine. But changing "addPlugin" so that it does not add anymore... Looks really strange.

**Comment by stevenzeck** [comment]
@mysticatea Can you post a link to the issue and/or PR where that change was done?  I get the feeling we're talking about different issues here.

**Comment by mysticatea** [comment]
https://github.com/eslint/eslint/issues/1971 is. The `addPlugin` was added to use specific plugin implementation instead of `node_modules/eslint-plugin-*`. It's never intended as an alternative to the `plugins` property.

**Comment by jehy** [comment]
May be it was not intended, but docs don't say anything about this replacement goal, and I think that this change will break many different tools...

**Comment by stevenzeck** [comment]
Yeah, to me this doesn't make much sense.  The point of `addPlugin()`, at least from the documentation, is to dynamically load a plugin to the CLIEngine without the need for the plugins property.  At the very least it should have been documented as a breaking change with v6 despite it not working as designed prior.

**Comment by platinumazure** [comment]
I think we need to look at how CLIEngine is meant to be used (from a design perspective) in these dynamic cases. Is CLIEngine intended to be a lightweight class that could, if needed, be instantiated on each lint run with different configurations? Or is it intended to be a stateful instance that should be modifiable dynamically? I'm not sure what the answer is right now, or what it should be.
